File: changji/src/changji/assembly/ffmpeg.py
"""FFmpeg 封装。

统一在这里调用，好处是外部依赖的检查只做一次，错误信息也能统一翻译成人话。

不用 MoviePy。它最新版停在 2025 年 5 月，一年多没有新版本，
项目说明里挂着招募维护者的告示，不适合压产线。
"""

from __future__ import annotations

import asyncio
import json
import shutil
from dataclasses import dataclass
from pathlib import Path


class FFmpegError(RuntimeError):
    pass


class FFmpegMissing(FFmpegError):
    """找不到 ffmpeg。装配环节的硬依赖。"""


@dataclass(frozen=True)
class MediaInfo:
    """一个媒体文件的关键信息。"""

    path: Path
    duration_s: float
    width: int = 0
    height: int = 0
    fps: float = 0.0
    frames: int = 0
    has_video: bool = False
    has_audio: bool = False
    pix_fmt: str = ""
    sar: str = ""

    @property
    def resolution(self) -> tuple[int, int]:
        return self.width, self.height


@dataclass(frozen=True)
class PixelStats:
    """一帧的亮度统计。用来判断画面是不是纯色或全黑。

    展布用 YHIGH 减 YLOW，也就是第 90 和第 10 百分位之差，而不是极差。
    极差会被单个亮点或暗点带偏：一张几乎全黑但有一个高光点的废图，
    极差能到 250，看起来很正常。百分位之差不吃这一套。
    """

    mean: float
    spread: float
    minimum: float
    maximum: float
    low: float = 0.0
    high: float = 0.0

    @property
    def looks_blank(self) -> bool:
        """近乎纯色。生成失败最常见的表现。

        阈值取 8。实测正常画面的展布在 40 以上，
        纯色或近乎纯色的画面在个位数。
        """
        return self.spread < 8.0

    @property
    def looks_clipped(self) -> bool:
        """整体过曝或全黑。"""
        return self.mean < 6.0 or self.mean > 249.0


class FFmpeg:
    """ffmpeg 与 ffprobe 的异步封装。"""

    def __init__(self, ffmpeg: str = "ffmpeg", ffprobe: str = "ffprobe") -> None:
        self.ffmpeg = ffmpeg
        self.ffprobe = ffprobe

    # ---- 可用性 ----

    def check(self) -> None:
        """启动时体检。缺了要在跑之前就说清楚，而不是跑到装配才炸。"""
        missing = [
            name for name, exe in (("ffmpeg", self.ffmpeg), ("ffprobe", self.ffprobe))
            if shutil.which(exe) is None
        ]
        if missing:
            raise FFmpegMissing(
                f"找不到 {'和'.join(missing)}。\n"
                f"Windows 可以用 winget install Gyan.FFmpeg，\n"
                f"macOS 用 brew install ffmpeg，\n"
                f"Linux 用包管理器安装。\n"
                f"装好后如果仍然找不到，在配置里填 assembly.ffmpeg_path 指定完整路径。"
            )

    def available(self) -> bool:
        try:
            self.check()
            return True
        except FFmpegMissing:
            return False

    # ---- 执行 ----

    async def _run(self, exe: str, args: list[str], timeout_s: float = 600.0) -> str:
        try:
            proc = await asyncio.create_subprocess_exec(
                exe, *args,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
            )
        except FileNotFoundError as exc:
            raise FFmpegMissing(f"找不到 {exe}") from exc
        try:
            out, err = await asyncio.wait_for(proc.communicate(), timeout=timeout_s)
        except asyncio.TimeoutError:
            proc.kill()
            raise FFmpegError(f"{exe} 执行超时（{timeout_s:.0f} 秒）") from None
        if proc.returncode != 0:
            tail = err.decode("utf-8", "replace").strip().splitlines()
            raise FFmpegError(
                f"{exe} 执行失败（退出码 {proc.returncode}）：\n"
                + "\n".join(tail[-8:])
            )
        return out.decode("utf-8", "replace")

    async def run_ffmpeg(self, args: list[str], timeout_s: float = 1800.0) -> str:
        return await self._run(self.ffmpeg, ["-hide_banner", "-nostdin", "-y", *args],
                               timeout_s)

    async def run_ffprobe(self, args: list[str], timeout_s: float = 120.0) -> str:
        return await self._run(self.ffprobe, ["-hide_banner", *args], timeout_s)

    # ---- 探测 ----

    async def probe(self, path: str | Path) -> MediaInfo:
        p = Path(path)
        if not p.is_file():
            raise FFmpegError(f"文件不存在：{p}")
        raw = await self.run_ffprobe([
            "-v", "error", "-print_format", "json",
            "-show_format", "-show_streams", str(p),
        ])
        try:
            data = json.loads(raw)
        except json.JSONDecodeError as exc:
            raise FFmpegError(f"ffprobe 输出无法解析：{p}") from exc

        streams = data.get("streams", [])
        video = next((s for s in streams if s.get("codec_type") == "video"), None)
        audio = next((s for s in streams if s.get("codec_type") == "audio"), None)

        duration = _to_float(data.get("format", {}).get("duration"))
        if duration == 0.0 and video:
            duration = _to_float(video.get("duration"))

        info = MediaInfo(
            path=p,
            duration_s=duration,
            has_video=video is not None,
            has_audio=audio is not None,
        )
        if video:
            info = MediaInfo(
                path=p,
                duration_s=duration,
                width=int(video.get("width") or 0),
                height=int(video.get("height") or 0),
                fps=_parse_fps(video.get("r_frame_rate", "0/1")),
                frames=int(video.get("nb_frames") or 0),
                has_video=True,
                has_audio=audio is not None,
                pix_fmt=video.get("pix_fmt", ""),
                sar=video.get("sample_aspect_ratio", ""),
            )
        return info

    async def pixel_stats(self, path: str | Path, at_second: float | None = None) -> PixelStats:
        """取一帧的亮度统计。

        用 ffmpeg 的 signalstats 滤镜算，不在 Python 里解码，
        省掉 numpy 和 opencv 依赖。
        """
        p = Path(path)
        args = []
        if at_second is not None:
            args += ["-ss", f"{at_second:.3f}"]
        args += [
            "-i", str(p), "-frames:v", "1",
            "-vf", "signalstats,metadata=print:file=-",
            "-f", "null", "-",
        ]
        # signalstats 把结果写到 stdout，失败时退回读 stderr
        out = await self.run_ffmpeg(args, timeout_s=120.0)
        return _parse_signalstats(out)

    async def sample_pixel_stats(
        self, path: str | Path, samples: int = 3,
    ) -> list[PixelStats]:
        """在片段的几个位置取样。只看一帧会漏掉中途崩坏的镜头。"""
        info = await self.probe(path)
        if info.duration_s <= 0:
            return []
        points = [
            info.duration_s * frac
            for frac in _sample_points(samples)
        ]
        out = []
        for t in points:
            try:
                out.append(await self.pixel_stats(path, at_second=t))
            except FFmpegError:
                continue
        return out

    async def measure_loudness(self, path: str | Path) -> dict[str, float]:
        """测响度。两遍归一法的第一遍。"""
        p = Path(path)
        args = [
            "-i", str(p), "-af",
            "loudnorm=I=-16:TP=-1.5:LRA=11:print_format=json",
            "-f", "null", "-",
        ]
        try:
            proc = await asyncio.create_subprocess_exec(
                self.ffmpeg, "-hide_banner", "-nostdin", *args,
                stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE,
            )
            _, err = await asyncio.wait_for(proc.communicate(), timeout=600)
        except FileNotFoundError as exc:
            raise FFmpegMissing(f"找不到 {self.ffmpeg}") from exc
        except asyncio.TimeoutError:
            raise FFmpegError("测响度超时") from None

        text = err.decode("utf-8", "replace")
        start = text.rfind("{")
        end = text.rfind("}")
        if start == -1 or end == -1:
            raise FFmpegError("loudnorm 没有返回测量结果，可能是文件里没有音轨")
        try:
            data = json.loads(text[start:end + 1])
        except json.JSONDecodeError as exc:
            raise FFmpegError(f"loudnorm 输出无法解析：{exc}") from exc
        return {k: _to_float(v) for k, v in data.items()
                if k.startswith(("input_", "output_", "target_"))}

    async def extract_frame(
        self, path: str | Path, dest: str | Path, at_second: float = 0.0,
    ) -> Path:
        """抽一帧存成图片。首尾帧比对和接龙都要用。"""
        target = Path(dest)
        target.parent.mkdir(parents=True, exist_ok=True)
        await self.run_ffmpeg([
            "-ss", f"{at_second:.3f}", "-i", str(path),
            "-frames:v", "1", "-q:v", "2", str(target),
        ], timeout_s=120.0)
        return target


def _to_float(value) -> float:
    try:
        return float(value)
    except (TypeError, ValueError):
        return 0.0


def _parse_fps(rate: str) -> float:
    try:
        num, _, den = rate.partition("/")
        d = float(den or 1)
        return float(num) / d if d else 0.0
    except ValueError:
        return 0.0


def _sample_points(samples: int) -> list[float]:
    """取样位置。避开首尾各 10%，那里常有编码伪影。"""
    if samples <= 1:
        return [0.5]
    lo, hi = 0.1, 0.9
    step = (hi - lo) / (samples - 1)
    return [lo + step * i for i in range(samples)]


_STAT_KEYS = {
    "lavfi.signalstats.YAVG": "mean",
    "lavfi.signalstats.YMIN": "minimum",
    "lavfi.signalstats.YMAX": "maximum",
    "lavfi.signalstats.YLOW": "low",
    "lavfi.signalstats.YHIGH": "high",
}


def _parse_signalstats(text: str) -> PixelStats:
    """解析 signalstats 的输出。

    signalstats 不输出标准差。YDIF 是相邻帧差异，单帧模式下根本不产生，
    拿它当标准差会让所有画面都被判成纯色。用 YLOW 和 YHIGH 这两个
    百分位数算展布才是对的。
    """
    found: dict[str, float] = {}
    for line in text.splitlines():
        key, sep, value = line.strip().partition("=")
        if not sep:
            continue
        name = _STAT_KEYS.get(key.strip())
        if name:
            found[name] = _to_float(value)

    if "mean" not in found:
        raise FFmpegError("signalstats 没有返回亮度统计，可能是文件损坏或没有视频轨")

    low = found.get("low", found.get("minimum", 0.0))
    high = found.get("high", found.get("maximum", 0.0))
    return PixelStats(
        mean=found.get("mean", 0.0),
        spread=max(0.0, high - low),
        minimum=found.get("minimum", 0.0),
        maximum=found.get("maximum", 0.0),
        low=low,
        high=high,
    )

File: changji/webapp/server/src/engine/ffmpeg.js
/**
 * ffmpeg 与 ffprobe 的封装。
 *
 * 解析那部分单独拆出来做成纯函数（parseProbe / parseSignalstats 等），
 * 不需要真的装 ffmpeg 就能测——而这些解析恰恰是最容易出错的地方：
 * 字段可能缺、可能是字符串、可能是 "0/0" 这种除不了的帧率。
 */

import { execFile } from 'node:child_process'
import fs from 'node:fs'

export class FFmpegError extends Error {}
export class FFmpegMissing extends FFmpegError {}

export const toFloat = (value) => {
  const n = Number(value)
  return Number.isFinite(n) ? n : 0
}

/** 帧率是 "24000/1001" 这种分数写法。分母是 0 时不能直接除。 */
export function parseFps(rate) {
  const [num, den] = String(rate ?? '').split('/')
  const d = Number(den || 1)
  if (!Number.isFinite(d) || d === 0) return 0
  const n = Number(num)
  return Number.isFinite(n) ? n / d : 0
}

/** 取样位置。避开首尾各 10%，那里常有编码伪影。 */
export function samplePoints(samples) {
  if (samples <= 1) return [0.5]
  const lo = 0.1
  const hi = 0.9
  const step = (hi - lo) / (samples - 1)
  return Array.from({ length: samples }, (_, i) => lo + step * i)
}

/** 一帧的亮度统计。 */
export const pixelStats = ({ mean = 0, spread = 0, minimum = 0, maximum = 0, low = 0, high = 0 }) => ({
  mean, spread, minimum, maximum, low, high,
})

const STAT_KEYS = {
  'lavfi.signalstats.YAVG': 'mean',
  'lavfi.signalstats.YMIN': 'minimum',
  'lavfi.signalstats.YMAX': 'maximum',
  'lavfi.signalstats.YLOW': 'low',
  'lavfi.signalstats.YHIGH': 'high',
}

/**
 * 解析 signalstats 的输出。
 *
 * signalstats 不输出标准差。YDIF 是相邻帧差异，单帧模式下根本不产生，
 * 拿它当标准差会让所有画面都被判成纯色——整条流水线会把每一个镜头
 * 都退回重跑，永远出不了片。用 YLOW 和 YHIGH 这两个百分位数算展布才对。
 */
export function parseSignalstats(text) {
  const found = {}
  for (const line of String(text).split('\n')) {
    const at = line.indexOf('=')
    if (at < 0) continue
    const name = STAT_KEYS[line.slice(0, at).trim()]
    if (name) found[name] = toFloat(line.slice(at + 1))
  }
  if (!('mean' in found)) {
    throw new FFmpegError('signalstats 没有返回亮度统计，可能是文件损坏或没有视频轨')
  }
  const low = found.low ?? found.minimum ?? 0
  const high = found.high ?? found.maximum ?? 0
  return pixelStats({
    mean: found.mean ?? 0,
    spread: Math.max(0, high - low),
    minimum: found.minimum ?? 0,
    maximum: found.maximum ?? 0,
    low,
    high,
  })
}

/** 解析 ffprobe 的 JSON 输出。 */
export function parseProbe(raw, path = '') {
  let data
  try {
    data = typeof raw === 'string' ? JSON.parse(raw) : raw
  } catch {
    throw new FFmpegError(`ffprobe 输出无法解析：${path}`)
  }

  const streams = data.streams ?? []
  const video = streams.find((s) => s.codec_type === 'video') ?? null
  const audio = streams.find((s) => s.codec_type === 'audio') ?? null

  // 有些封装格式的时长只写在流上，不写在容器上
  let duration = toFloat(data.format?.duration)
  if (duration === 0 && video) duration = toFloat(video.duration)

  return {
    path,
    duration_s: duration,
    width: video ? Number(video.width) || 0 : 0,
    height: video ? Number(video.height) || 0 : 0,
    fps: video ? parseFps(video.r_frame_rate ?? '0/1') : 0,
    frames: video ? Number(video.nb_frames) || 0 : 0,
    has_video: video !== null,
    has_audio: audio !== null,
    pix_fmt: video?.pix_fmt ?? '',
    sar: video?.sample_aspect_ratio ?? '',
  }
}

/** 解析 loudnorm 第一遍的输出。两遍归一法要拿它当第二遍的输入。 */
export function parseLoudnorm(text) {
  // loudnorm 把 JSON 打在 stderr 末尾，前面还有一堆日志
  const at = String(text).lastIndexOf('{')
  if (at < 0) throw new FFmpegError('loudnorm 没有返回测量结果')
  try {
    const data = JSON.parse(String(text).slice(at))
    return {
      input_i: toFloat(data.input_i),
      input_tp: toFloat(data.input_tp),
      input_lra: toFloat(data.input_lra),
      input_thresh: toFloat(data.input_thresh),
      target_offset: toFloat(data.target_offset),
    }
  } catch {
    throw new FFmpegError('loudnorm 的测量结果不是合法 JSON')
  }
}

export class FFmpeg {
  constructor({ ffmpeg = 'ffmpeg', ffprobe = 'ffprobe', execImpl = null } = {}) {
    this.ffmpeg = ffmpeg
    this.ffprobe = ffprobe
    this.exec = execImpl ?? defaultExec
  }

  /**
   * 启动时体检。
   *
   * 缺了要在跑之前就说清楚，而不是跑到装配那一步才炸——那时候几十分钟的
   * 渲染已经跑完了。
   */
  async check() {
    const missing = []
    for (const [name, exe] of [['ffmpeg', this.ffmpeg], ['ffprobe', this.ffprobe]]) {
      try {
        await this.exec(exe, ['-version'], 15000)
      } catch {
        missing.push(name)
      }
    }
    if (missing.length) {
      throw new FFmpegMissing(
        `找不到 ${missing.join(' 和 ')}。装配成片要用它们，` +
          `装好之后确认在 PATH 里，或者去设置里填绝对路径。`,
      )
    }
  }

  async available() {
    try {
      await this.check()
      return true
    } catch {
      return false
    }
  }

  runFfmpeg(args, timeoutMs = 1800000) {
    return this.exec(this.ffmpeg, ['-hide_banner', '-nostdin', '-y', ...args], timeoutMs)
  }

  runFfprobe(args, timeoutMs = 120000) {
    return this.exec(this.ffprobe, ['-hide_banner', ...args], timeoutMs)
  }

  async probe(file) {
    if (!fs.existsSync(file)) throw new FFmpegError(`文件不存在：${file}`)
    const { stdout } = await this.runFfprobe([
      '-v', 'error', '-print_format', 'json', '-show_format', '-show_streams', String(file),
    ])
    return parseProbe(stdout, String(file))
  }

  /**
   * 取一帧的亮度统计。
   *
   * 用 ffmpeg 的 signalstats 滤镜算，不在 Node 里解码——那样要拉一个
   * 解码库进来，而这一步只需要几个数字。
   */
  async pixelStats(file, atSecond = null) {
    const args = []
    if (atSecond !== null) args.push('-ss', atSecond.toFixed(3))
    args.push(
      '-i', String(file), '-frames:v', '1',
      '-vf', 'signalstats,metadata=print:file=-',
      '-f', 'null', '-',
    )
    const { stdout, stderr } = await this.runFfmpeg(args, 120000)
    // signalstats 把结果写到 stdout，失败时退回读 stderr
    return parseSignalstats(stdout || stderr)
  }

  /** 在片段的几个位置取样。只看一帧会漏掉中途崩坏的镜头。 */
  async samplePixelStats(file, samples = 3) {
    const info = await this.probe(file)
    if (info.duration_s <= 0) return []
    const out = []
    for (const frac of samplePoints(samples)) {
      try {
        // eslint-disable-next-line no-await-in-loop
        out.push(await this.pixelStats(file, info.duration_s * frac))
      } catch {
        // 某一个取样点读不出来不该让整个镜头判不了，跳过就是
      }
    }
    return out
  }

  /** 测响度。两遍归一法的第一遍。 */
  async measureLoudness(file, target) {
    const filter =
      `loudnorm=I=${target.i}:TP=${target.tp}:LRA=${target.lra}:print_format=json`
    const { stdout, stderr } = await this.runFfmpeg(
      ['-i', String(file), '-af', filter, '-f', 'null', '-'],
      600000,
    )
    return parseLoudnorm(stderr || stdout)
  }
}

function defaultExec(exe, args, timeoutMs) {
  return new Promise((resolve, reject) => {
    execFile(
      exe,
      args,
      { timeout: timeoutMs, maxBuffer: 32 * 1024 * 1024, encoding: 'utf8' },
      (err, stdout, stderr) => {
        if (!err) return resolve({ stdout, stderr })
        if (err.code === 'ENOENT') return reject(new FFmpegMissing(`找不到 ${exe}`))
        if (err.killed) {
          return reject(new FFmpegError(`${exe} 执行超时（${(timeoutMs / 1000).toFixed(0)} 秒）`))
        }
        // 只留最后几行。ffmpeg 的日志很长，前面全是编码参数，
        // 真正的原因在最后
        const tail = String(stderr ?? '').trim().split('\n').slice(-8).join('\n')
        return reject(
          new FFmpegError(`${exe} 执行失败（退出码 ${err.code}）：\n${tail}`),
        )
      },
    )
  })
}
